sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java

```diff
@@ -551,9 +548,9 @@ public static TableRow toTableRow(Row row) {
         return toTableRow((Row) fieldValue);
 
       case DATETIME:
-        return ((Instant) fieldValue)
-            .toDateTime(DateTimeZone.UTC)
-            .toString(BIGQUERY_TIMESTAMP_PRINTER);
+        org.joda.time.Instant jodaInstant = (org.joda.time.Instant) fieldValue;
+        java.time.Instant javaInstant = java.time.Instant.ofEpochMilli(jodaInstant.getMillis());
+        return BIGQUERY_TIMESTAMP_PRINTER.format(javaInstant);
```

Comment:
@pabloem @TheNeuralBit This part was missing from my earlier attempt as I did not fully understand the codebase at that point in time. The TableRow just contained a String in the Java Instant format, and not the native BigQuery TIMESTAMP format as it was skipping past all the if statements. I apologize for this slip-up.

As for whether to reject or truncate nanos to micros, I think we should follow the pattern established by other SQL types and do the truncation implicitly. Anyway, this commit will be an improvement as the previous SQL DATETIME used org.joda.time.Instant, which is limited to microseconds precision. The switch to java.time.Instant will improve it to nanos precision, and then truncating to micros at the last step when formatting to String for TableRow to accept.

Regarding @TheNeuralBit suggestion to allow the option for selecting, I think it will increase the scope of the ticket, which is originally to solve the Protobuf -> BigQuery pipeline for google.protobuf.Timestamp fields. We can open another ticket to figure out what options are open to user configuration, and how best to pass it to the existing control flow.

Reply:
(FYI I edited your message, you said DATETIME/joda Instant was limited to microseconds, but I think you meant milliseconds)

You're right this is an improvement over the existing solution (converting NanosInstant to DATETIME with millisecond precision), because it gets the full supported microsecond precision into BigQuery. But there is an advantage with the existing solution: the user has to opt-in to the truncation, by converting to DATETIME. I feel strongly that we shouldn't be truncating implicitly, and if we're doing that elsewhere I'd consider it a bug. Perhaps another reviewer would feel differently, but that is my position.

That being said, you make a fair point that adding an option may be more work than you want to take on. Would it be a reasonable compromise to make this PR use the "erroring" approach, and file a follow-up ticket to add an option for truncation?

Reply:
Thanks @TheNeuralBit . I'm working together with @amuletxheart and also looking to see if I can be of any help.

With the opt-in flag, would this below approach be a reasonable venue to proceed?

- add a `BigQueryIO.Write#allowTruncatedTimestamps()` method for explicit opt-in (i.e. default false)
- pass the value to `BigQueryUtils.toTableRow()` as another parameter at the call site [here](https://github.com/apache/beam/blob/0111cff88025f0dc783a0890078b769139c8ae36/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L2687), and `BigQueryUtils.toTableSchema()` [here](https://github.com/apache/beam/blob/0111cff88025f0dc783a0890078b769139c8ae36/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java#L2690)
- Update `BigQueryUtils.toTableSchema()` to only accept NanosInstant logical type if the `allowTruncatedTimestamps` parameter passed in is true. This should reject, i.e. error out, before any rows are processed, i.e. before formatFunction is triggered. In fact, with this, I'm thinking we don't even need to pass `allowTruncatedTimestamps` to `BigQueryUtils.toTableRow()` any more.

Please let us know what do you think? @TheNeuralBit and other maintainers. Thanks a lot!

Reply:
It's a little messy that we'd need to pass the option as an argument in `toTableRow`, but I think to do anything cleaner would take a major re-architecting. We might create another type like ConversionOptions for this. 

I have an important note regarding this statement though:
> This should reject, i.e. error out, before any rows are processed, i.e. before formatFunction is triggered.

It shouldn't error out universally. Instead it should raise an error _if the timestamp requires nanosecond precision_. That is, the timestamp 0.123456000 would be processed fine, since we can truncate it to microsecond precision without losing any information. However, the timestamp 0.123456789 would yield an error at execution time.

Similar to this logic from the read side: https://github.com/apache/beam/blob/c8568eb372e2e432ce0a9998fa2864c090ddf44a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryUtils.java#L762-L765

Reply:
Got it. IIUC, this means we should not error out in `BigQueryUtils.toTableSchema()` but rather check the flag and error out case by case in `BigQueryUtils.toTableRow()` as the snippet you linked.

I think @amuletxheart and I will have to defer the re-architecting work to the maintainers who definitely have a much better understanding of the overall architecture than us, even though I agree passing an argument to `toTableRow()` is a bit messy.

That said, are you okay with the proposed `BigQueryIO.Write#allowTruncatedTimestamps()` change, as I think this is kinda part of the "public API" of the BigQueryIO, while `toTableRow()` is an private implementation that is easier to change when we later undergo the re-architecturing work.

Reply:
Sorry forgot to tag @TheNeuralBit in my previous reply. Edited.

Reply:
Yes `BigQueryIO.Write#allowTruncatedTimestamps()` sounds like a reasonable addition to the public API of BigQuery.  I might suggest `withAllowTruncatedTimestamps()` to be consistent with other boolean parameters, e.g. [`withExtendedErrorInfo`](https://beam.apache.org/releases/javadoc/2.32.0/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.Write.html#withExtendedErrorInfo--). @chamikaramj may be a better judge of what the public API should look like though.

The reason I clarified my suggested default (non-truncating, error-raising) behavior was just in case that behavior is acceptable for your use-case. If it is then you could avoid creating the parameter for now if you want.